File: CelulaReport/app/src/main/java/com/example/celulareport/ui/fragment/ReportsListFragment.java
package com.example.celulareport.ui.fragment;


import android.graphics.Color;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ActionMode;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.example.celulareport.R;
import com.example.celulareport.util.Constraint;
import com.example.celulareport.ui.adapter.CardAdapter;
import com.example.celulareport.viewmodel.ReportListViewModel;

import java.util.List;

import static androidx.appcompat.view.ActionMode.*;

public class ReportsListFragment extends Fragment implements CardAdapter.OnCardClickListener, CardAdapter.OnCardLongCLickListener {

    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_MONTH = "Month";

    private String mMonth;

    private RecyclerView mRecycler;

    private CardAdapter mAdapter;

    private CardView mCardView;

    private ReportListViewModel mViewModel;

    private ProgressBar progressBar;

    // Get position item when long click is activate
    int longClickPosition;

    //Context menu mode
    private ActionMode mActionMode;

    public ReportsListFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param selectMonth month selected in list of month of report.
     * @return A new instance of fragment ReportsListFragment.
     */
    public static ReportsListFragment newInstance(String selectMonth) {
        ReportsListFragment fragment = new ReportsListFragment();
        Bundle args = new Bundle();
        args.putString(ARG_MONTH, selectMonth);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        if (getArguments() != null) {
            mMonth = getArguments().getString(ARG_MONTH);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View v = inflater.inflate(R.layout.fragment_reports_list, container, false);

        mRecycler = v.findViewById(R.id.rv_report_list);

        //get Progress Bar
        progressBar = v.findViewById(R.id.progress_reports);

        //Setup toolbar as actionbar
        SetupToolbar(mMonth, v);
        return v;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        //Create view model to share between fragments
        mViewModel =  new ViewModelProvider(requireActivity()).get(ReportListViewModel.class);

        //Observe if there're changes in reports Card list
        subscribeReportList(mViewModel, mMonth);

        //initializing Recyclerview
        SetupRecycler();
    }

    //Setup toolbar and collapsing bar
    private void SetupToolbar(String title, @NonNull View v){
        Toolbar mToolbar = (Toolbar) v.findViewById(R.id.reports_toolbar);
        TextView textTitle = (TextView) v.findViewById(R.id.month_title);
        textTitle.setText(title);
        mToolbar.setLogo(R.drawable.ic_report);
        ((AppCompatActivity)requireActivity()).setSupportActionBar(mToolbar);
        //ActionBar actionBar = ((AppCompatActivity)getActivity()).getSupportActionBar();
        //actionBar.setDisplayShowTitleEnabled(false);
        //actionBar.setDisplayHomeAsUpEnabled(true);
    }

    //Adding recycleView(list of reports)
    public void SetupRecycler(){

        final int nColumn = 2;
        GridLayoutManager layoutManager = new GridLayoutManager(getContext(), nColumn);
        mRecycler.setLayoutManager(layoutManager);
        mAdapter = new CardAdapter(getContext(), this, this);
        mRecycler.setAdapter(mAdapter);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.main_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
        MenuItem item = menu.findItem(R.id.ic_reports_search);
        SearchView searchView = (SearchView) item.getActionView();


        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {

                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                progressBar.setVisibility(View.VISIBLE);
                new Thread(){
                    @Override
                    public void run() {
                        mAdapter.getFilter().filter(newText);
                        requireActivity().runOnUiThread(()->progressBar.setVisibility(View.GONE));
                    }
                }.start();

                return false;
            }
        });
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

            if(item.getItemId() == R.id.ic_add_report){
                //Toast.makeText(getContext(), "Click in add report icon!", Toast.LENGTH_SHORT).show();
                AddReportFragment fragment = new AddReportFragment();
                requireActivity()
                        .getSupportFragmentManager()
                        .beginTransaction()
                        .setReorderingAllowed(true)
                        .setCustomAnimations(R.anim.enter_slide_up, R.anim.exit_fade_freeze, R.anim.enter_fade_freeze, R.anim.exit_slide_down)
                        .replace(R.id.main_container, fragment, fragment.getClass().getName())
                        .addToBackStack(null)
                        .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                        .commit();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    // menu context when we have a long click in the card
    public Callback getActionModeCallBack(){


        return new Callback() {
            @Override
            public boolean onCreateActionMode(@NonNull ActionMode mode, Menu menu) {
                mode.getMenuInflater().inflate(R.menu.card_context_menu, menu);
                mode.setTitle("opções");
                return true;
            }

            @Override
            public boolean onPrepareActionMode(ActionMode mode, Menu menu) {

                return false;
            }

            @Override
            public boolean onActionItemClicked(ActionMode mode, @NonNull MenuItem item) {

                switch (item.getItemId()){
                    case R.id.ic_delete_card:
                        long id = mAdapter.getReportId(longClickPosition);
                        mViewModel.deleteById(id);
                        mode.finish();
                        return true;

                    case R.id.ic_share_card:
                        Toast.makeText(getContext(), "Shared icon is still Disabled.", Toast.LENGTH_LONG).show();
                        mode.finish();
                        return true;

                    default:
                        return false;
                }
            }

            @Override
            public void onDestroyActionMode(ActionMode mode) {

                mCardView.setCardBackgroundColor(Color.parseColor("#455A64"));
                mCardView = null;
                mActionMode = null;
            }
        };
    }

    @Override
    public void onCardClick(int position, CardView mCardView) {
        if(mActionMode != null){
            mActionMode.finish();
        }
        else{
            //Toast.makeText(getContext(), "Clicked item " + position, Toast.LENGTH_SHORT).show();
            long reportID;
            reportID = mAdapter.getReportId(position);

            ReportDetailsFragment fragment = ReportDetailsFragment.newInstance(reportID);
            requireActivity()
                    .getSupportFragmentManager()
                    .beginTransaction()
                    .setReorderingAllowed(true)
                    .setCustomAnimations(R.anim.enter_fade_in,
                            R.anim.exit_fade_in,
                            R.anim.enter_fade_in,
                            R.anim.exit_fade_in)
                    .replace(R.id.main_container, fragment, fragment.getClass().getName())
                    .addToBackStack(null)
                    .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                    .commit();
        }
    }

    @Override
    public boolean onCardLongClick(int position, CardView mCardView) {

        if (mActionMode == null){

            //trigger menu context
            Callback mCallback = getActionModeCallBack();
            mActionMode = ((AppCompatActivity)requireActivity()).startSupportActionMode(mCallback);
            longClickPosition = position;
            this.mCardView = mCardView;
            this.mCardView.setCardBackgroundColor(Color.parseColor("#37474F"));
            this.mCardView.setSelected(true);



        }else {

            mCardView.setSelected(false);
            mActionMode.finish();
        }


        return true;
    }

    //Verify the month to return number according to month to search in database
    private String returnMonthValue(String monthSelect){
        String monthValue = "";
        switch (monthSelect){
            case Constraint.STRING_JANUARY:
                return monthValue = Constraint.VALUE_JANUARY;

            case Constraint.STRING_FEBRUARY:
                return monthValue = Constraint.VALUE_FEBRUARY;

            case Constraint.STRING_MARCH:
                return monthValue = Constraint.VALUE_MARCH;

            case Constraint.STRING_APRIL:
                return monthValue = Constraint.VALUE_APRIL;

            case Constraint.STRING_MAY:
                return monthValue = Constraint.VALUE_MAY;

            case Constraint.STRING_JUNE:
                return monthValue = Constraint.VALUE_JUNE;

            case Constraint.STRING_JULY:
                return monthValue = Constraint.VALUE_JULY;

            case Constraint.STRING_AUGUST:
                return monthValue = Constraint.VALUE_AUGUST;

            case Constraint.STRING_SEPTEMBER:
                return monthValue = Constraint.VALUE_SEPTEMBER;

            case Constraint.STRING_OCTOBER:
                return monthValue = Constraint.VALUE_OCTOBER;

            case Constraint.STRING_NOVEMBER:
                return monthValue = Constraint.VALUE_NOVEMBER;

            case Constraint.STRING_DECEMBER:
                return monthValue = Constraint.VALUE_DECEMBER;

            default:
                break;
        }

        return monthValue;
    }

    //observe changes in report cards list
    public void subscribeReportList(ReportListViewModel viewModel, String monthSelect){

        try {
            //Verify month value to query by it
            String monthValue = returnMonthValue(monthSelect);
            //select reports by month and observe changes in that
            progressBar.setVisibility(View.VISIBLE);
            viewModel.getReportsByMonth(monthValue).observe(getViewLifecycleOwner(), reportCards -> {
                new Thread(){
                    @Override
                    public void run() {

                        //a few delay
                        try {
                            Thread.sleep(500);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                        //change ui state
                        requireActivity().runOnUiThread(()->{
                            mAdapter.setReportCardsByMonth(reportCards);
                            progressBar.setVisibility(View.GONE);});
                    }
                }.start();
            });

        } catch (NullPointerException exception){
            exception.printStackTrace();
        }
    }

}

File: CelulaReport/app/src/main/java/com/example/celulareport/ui/fragment/ReportsListFragment.kt
package com.example.celulareport.ui.fragment

import android.annotation.SuppressLint
import android.graphics.Color
import android.os.Bundle
import android.view.*
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ActionMode
import androidx.appcompat.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.cardview.widget.CardView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentTransaction
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.celulareport.R
import com.example.celulareport.db.model.ReportCard
import com.example.celulareport.ui.Constraint
import com.example.celulareport.ui.adapter.CardAdapter
import com.example.celulareport.ui.adapter.CardAdapter.OnCardClickListener
import com.example.celulareport.ui.adapter.CardAdapter.OnCardLongCLickListener
import com.example.celulareport.viewmodel.ReportListViewModel

/**
 * [ReportsListFragment] is fragment class to show up report card list
 *
 *
 */
class ReportsListFragment : Fragment(), OnCardClickListener, OnCardLongCLickListener {
    private var mMonth: String? = null
    private var mRecycler: RecyclerView? = null
    private var mAdapter: CardAdapter? = null
    private var mToolbar: Toolbar? = null
    private var textTitle: TextView? = null
    private var mCardView: CardView? = null
    private var mViewModel: ReportListViewModel? = null

    // Get position item when long click is activate
    var longClickPosition = 0

    //Context menu mode
    private var mActionMode: ActionMode? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        if (arguments != null) {
            mMonth = arguments!!.getString(ARG_MONTH)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        val v = inflater.inflate(R.layout.fragment_reports_list, container, false)
        mRecycler = v.findViewById(R.id.rv_report_list)

        //Setup toolbar as actionbar
        SetupToolbar(mMonth, v)
        return v
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        //Create view model to share between fragments
        mViewModel = ViewModelProvider(requireActivity()).get(
            ReportListViewModel::class.java
        )

        //Observe if there're changes in reports Card list
        subscribeReportList(mViewModel!!, mMonth)

        //initializing Recyclerview
        SetupRecycler()
    }

    //Setup toolbar and collapsing bar
    @SuppressLint("RestrictedApi")
    private fun SetupToolbar(title: String?, v: View) {
        mToolbar = v.findViewById<View>(R.id.reports_toolbar) as Toolbar
        textTitle = v.findViewById<View>(R.id.month_title) as TextView
        textTitle!!.text = title
        mToolbar!!.setLogo(R.drawable.ic_report)
        (activity as AppCompatActivity?)!!.setSupportActionBar(mToolbar)
        //ActionBar actionBar = ((AppCompatActivity)getActivity()).getSupportActionBar();
        //actionBar.setDisplayShowTitleEnabled(false);
        //actionBar.setDisplayHomeAsUpEnabled(true);
    }

    //Adding recycleView(list of reports)
    fun SetupRecycler() {
        val nColumn = 2
        val layoutManager = GridLayoutManager(context, nColumn)
        mRecycler!!.layoutManager = layoutManager
        mAdapter = CardAdapter(context, this, this)
        mRecycler!!.adapter = mAdapter
        //divisor between cards
        //mRecycler.addItemDecoration(new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL));
        //mRecycler.addItemDecoration(new DividerItemDecoration(getContext(), DividerItemDecoration.HORIZONTAL));
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.main_menu, menu)
        super.onCreateOptionsMenu(menu, inflater)
        val item = menu.findItem(R.id.ic_reports_search)
        val searchView = item.actionView as SearchView
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                mAdapter!!.filter.filter(newText)
                return false
            }
        })
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.ic_add_report -> {
                //Toast.makeText(getContext(), "Click in add report icon!", Toast.LENGTH_SHORT).show();
                val fragment = AddReportFragment()
                activity
                    .getSupportFragmentManager()
                    .beginTransaction()
                    .setReorderingAllowed(true)
                    .setCustomAnimations(
                        R.anim.enter_slide_up,
                        R.anim.exit_fade_freeze,
                        R.anim.enter_fade_freeze,
                        R.anim.exit_slide_down
                    )
                    .replace(R.id.main_container, fragment, fragment.javaClass.name)
                    .addToBackStack(null)
                    .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                    .commit()
                return true
            }
            else -> {
            }
        }
        return false
    }

    // menu context when we have a long click in the card
    val actionModeCallBack: ActionMode.Callback
        get() = object : ActionMode.Callback {
            override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
                mode.menuInflater.inflate(R.menu.card_context_menu, menu)
                mode.title = "opções"
                return true
            }

            override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
                return false
            }

            override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
                return when (item.itemId) {
                    R.id.ic_delete_card -> {
                        val id = mAdapter!!.getReportId(longClickPosition)
                        mViewModel.deleteById(id)
                        mode.finish()
                        true
                    }
                    R.id.ic_share_card -> {
                        Toast.makeText(
                            context,
                            "Shared icon is still Disabled.",
                            Toast.LENGTH_LONG
                        ).show()
                        mode.finish()
                        true
                    }
                    else -> false
                }
            }

            override fun onDestroyActionMode(mode: ActionMode) {
                mCardView!!.setCardBackgroundColor(Color.parseColor("#455A64"))
                mCardView = null
                mActionMode = null
            }
        }

    override fun onCardClick(position: Int, mCardView: CardView) {
        if (mActionMode != null) {
            mActionMode!!.finish()
        } else {
            //Toast.makeText(getContext(), "Clicked item " + position, Toast.LENGTH_SHORT).show();
            val reportID: Long
            reportID = mAdapter!!.getReportId(position)
            val fragment = ReportDetailsFragment.newInstance(reportID)
            activity
                .getSupportFragmentManager()
                .beginTransaction()
                .setReorderingAllowed(true)
                .setCustomAnimations(
                    R.anim.enter_fade_in,
                    R.anim.exit_fade_in,
                    R.anim.enter_fade_in,
                    R.anim.exit_fade_in
                )
                .replace(R.id.main_container, fragment, fragment.javaClass.name)
                .addToBackStack(null)
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
                .commit()
        }
    }

    @SuppressLint("ResourceAsColor")
    override fun onCardLongClick(position: Int, mCardView: CardView): Boolean {
        if (mActionMode == null) {

            //trigger menu context
            val mCallback = actionModeCallBack
            mActionMode = (activity as AppCompatActivity?)!!.startSupportActionMode(mCallback)
            longClickPosition = position
            this.mCardView = mCardView
            this.mCardView!!.setCardBackgroundColor(Color.parseColor("#37474F"))
            this.mCardView!!.isSelected = true
        } else {
            mCardView.isSelected = false
            mActionMode!!.finish()
        }
        return true
    }

    //Verify the month to return number according to month to search in database
    private fun returnMonthValue(monthSelect: String?): String {
        var monthValue: String? = ""
        return when (monthSelect) {
            Constraint.STRING_JANUARY -> Constraint.VALUE_JANUARY.also { monthValue = it }
            Constraint.STRING_FEBRUARY -> Constraint.VALUE_FEBRUARY.also { monthValue = it }
            Constraint.STRING_MARCH -> Constraint.VALUE_MARCH.also { monthValue = it }
            Constraint.STRING_APRIL -> Constraint.VALUE_APRIL.also { monthValue = it }
            Constraint.STRING_MAY -> Constraint.VALUE_MAY.also { monthValue = it }
            Constraint.STRING_JUNE -> Constraint.VALUE_JUNE.also { monthValue = it }
            Constraint.STRING_JULY -> Constraint.VALUE_JULY.also { monthValue = it }
            Constraint.STRING_AUGUST -> Constraint.VALUE_AUGUST.also { monthValue = it }
            Constraint.STRING_SEPTEMBER -> Constraint.VALUE_SEPTEMBER.also { monthValue = it }
            Constraint.STRING_OCTOBER -> Constraint.VALUE_OCTOBER.also { monthValue = it }
            Constraint.STRING_NOVEMBER -> Constraint.VALUE_NOVEMBER.also { monthValue = it }
            Constraint.STRING_DECEMBER -> Constraint.VALUE_DECEMBER.also { monthValue = it }
            else -> throw NullPointerException("Value don't exist!")
        }
    }

    //observe changes in report cards list
    fun subscribeReportList(viewModel: ReportListViewModel, monthSelect: String?) {
        try {

            //Verify month value to query by it
            val monthValue = returnMonthValue(monthSelect)
            //select reports by month and observe changes in that
            viewModel.getReportsByMonth(monthValue).observe(
                viewLifecycleOwner,
                { reportCards: List<ReportCard?>? -> mAdapter!!.setReportCardsByMonth(reportCards) })
        } catch (exception: NullPointerException) {
            Toast.makeText(context, exception.message.toString(), Toast.LENGTH_LONG)
        }
    }

    companion object {
        // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
        private const val ARG_MONTH = "Month"

        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param selectMonth month selected in list of month of report.
         * @return A new instance of fragment ReportsListFragment.
         */
        @JvmStatic
        fun newInstance(selectMonth: String?): ReportsListFragment {
            val fragment = ReportsListFragment()
            val args = Bundle()
            args.putString(ARG_MONTH, selectMonth)
            fragment.arguments = args
            return fragment
        }
    }
}